### ISSUE TITLE
GRPO - Multiple Policy updates for one completion is not desired

### ISSUE THREAD

**Issue by SonuDixit** [huggingface/trl#3197]
Hi Team,
[Here](https://github.com/huggingface/trl/blob/c5ba70d4fc7cfc2e4515654eb03332d0ff4704d9/trl/trainer/grpo_trainer.py#L571), num_iterations($\mu$) optimisation steps are made on the same completion(trajectory). Please see the attached image from PPO paper - 
<img width="1027" alt="Screenshot 2025-03-31 at 6 15 44 PM" src="https://github.com/user-attachments/assets/1eb2524d-86a4-424e-8c98-1b17804ed80b" />

Attaching algorithm from DeepseekMath paper - 
![Screenshot 2025-03-31 at 6 17 50 PM](https://github.com/user-attachments/assets/b3113f34-1d43-43db-ab93-f68e2195a0b5)
I think, for line 11 in the above Algorithm - a different batch size $B$ is used such that $\mu * B \approx |D_b|$. A total of $\mu$ updates are being done using the batch size $B$. 

Please see the PPO algorithm - 
<img width="1044" alt="Screenshot 2025-03-31 at 6 44 42 PM" src="https://github.com/user-attachments/assets/9c331882-2636-4fd0-a7a8-8e9e56546548" />
I think $k$ in the PPO-Algo corresponds to $\mu$ in the GRPO. PPO Authors have mentioned hyperparams in appendix A, wherein it can be seen that K*M is less than or approximately equal to NT.

**Comment by qgallouedec** [comment]
That's an interesting point. So, you're saying that we should think of μ as the number of mini-batches the batch is divided into, rather than as the number of times the same batch is reused?
By the way, have you observed what is described in the paper “it often leads to destructively large policy updates”? I'm curious to know if the same thing happens with GRPO.

**Comment by SonuDixit** [comment]
“it often leads to destructively large policy updates” - I have not tried it with GRPO, but I think, it's true for any on-policy Policy Gradient Algo. I can think of two reasons - 1) overfitting to a batch, and 2) once the policy has been updated the previous batch trajectories are no longer from the policy we want to update. Since PPO(hence GRPO) objective is a lower bound on the unclipped PG objective $log \pi_\theta(a_t|s_t)A_t$, the condition can be slightly relaxed, It won't hurt as much if $KM$ is slightly higher than $NT$. ref - PPO paper-Appendix, Table4-Roboschool_Locomotion.

In GRPO, all the tokens get the same advantage as the last token. It is much more aggressive in pushing good trajectory prob high, and lowering the bad trajectory prob. I think, here we need to be much more strict in keeping $KM \approx NT$. From DeepSeekMath-5.2.2 - "it seems that the improvement is attributed to **boosting the correct response from TopK** rather than the enhancement of fundamental capabilities." If we fit $\mu$ times on the same batch, we might be overfitting to the correct response. 

I think, we should train on $\mu$ mini-batches from $D_b$.

**Comment by avishaiElmakies** [comment]
@qgallouedec 
I was wondering about this as well, is this being handled or something?
I did see that in my case using `num_iterations > 1` gives much worse results than using `num_iterations == 1`.  Which was kinda of weird to me since it makes the clipping of PPO/GRPO irrelevant. for PPO the clipping was needed and without it the model will be very bad.
I also believe this is how verl grpo behaves https://verl.readthedocs.io/en/latest/algo/grpo.html as seen with the argument `actor_rollout_ref.actor.ppo_mini_batch_size`. they also have `ppo_epochs` which seems to be similar to ours `num_iterations`.
So i was wondering if a change to how `num_iterations` behaves is being planned.
Maybe we don't have to change `num_iterations` and we can just add `ppo_mini_batch_size` as another option?

**Comment by avishaiElmakies** [comment]
Ok, after looking into this more. it seems i missed `steps_per_generation` which seems to do the `mini_batch_size` thing i want.
I will admit that the documentation for this argument is not very clear

**Comment by qgallouedec** [comment]
Thanks for the feedback. The doc could be clearer. If you think of way to make it clearer, feel free to suggest.

**Comment by avishaiElmakies** [comment]
I will try to think of what can be done, and open a PR if i have an idea.
Thanks!

**Comment by albertvillanova** [comment]
Hey @SonuDixit, sorry for the very slow reply — we're catching up on the backlog. 

The implementation is actually faithful to the paper: `num_iterations` (μ) is the number of PPO-style epochs, i.e. how many times the *same* generation batch is reused, and that reuse is made principled by importance sampling + clipping (TRL stores the old per-token log-probs whenever μ>1). The behavior you describe — splitting D_b into mini-batches with one update each and no reuse — is exactly the *default*: with `num_iterations=1` nothing is reused, and `steps_per_generation` controls how a generation batch is split into distinct mini-batch updates (see https://github.com/huggingface/trl/blob/b58947c7e3f6a3ccddc56771f4bfa65c0d72bb41/trl/trainer/grpo_trainer.py#L935-L944 and the `steps_per_generation` doc at https://github.com/huggingface/trl/blob/b58947c7e3f6a3ccddc56771f4bfa65c0d72bb41/trl/trainer/grpo_config.py#L84-L86). So there's nothing to fix in the algorithm itself — but as @avishaiElmakies noted, the naming/docs are genuinely confusing, so a doc-clarification PR would be very welcome. 

Closing as intended behavior; happy to reopen if anything still looks off.
